libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/DialogManager.cs

```diff
@@ -149,21 +150,40 @@ public async Task<DialogManagerResult> OnTurnAsync(ITurnContext context, Cancell
             await dc.GetState().LoadAllScopesAsync(cancellationToken).ConfigureAwait(false);
 
             DialogTurnResult turnResult = null;
-            if (dc.ActiveDialog == null)
-            {
-                // start root dialog
-                turnResult = await dc.BeginDialogAsync(this.rootDialogId, cancellationToken: cancellationToken).ConfigureAwait(false);
-            }
-            else
-            {
-                // Continue execution
-                // - This will apply any queued up interruptions and execute the current/next step(s).
-                turnResult = await dc.ContinueDialogAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
 
-                if (turnResult.Status == DialogTurnStatus.Empty)
+            while (true)
+            {
+                try
                 {
-                    // restart root dialog
-                    turnResult = await dc.BeginDialogAsync(this.rootDialogId, cancellationToken: cancellationToken).ConfigureAwait(false);
+                    if (dc.ActiveDialog == null)
+                    {
+                        // start root dialog
+                        turnResult = await dc.BeginDialogAsync(this.rootDialogId, cancellationToken: cancellationToken).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        // Continue execution
+                        // - This will apply any queued up interruptions and execute the current/next step(s).
+                        turnResult = await dc.ContinueDialogAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
+
+                        if (turnResult.Status == DialogTurnStatus.Empty)
+                        {
+                            // restart root dialog
+                            turnResult = await dc.BeginDialogAsync(this.rootDialogId, cancellationToken: cancellationToken).ConfigureAwait(false);
+                        }
+                    }
+
+                    break;
+                }
+                catch (Exception err)
+                {
+                    // fire error event, bubbling from the leaf.
+                    var handled = await dc.EmitEventAsync(DialogEvents.Error, err, bubble: true, fromLeaf: true, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
You don't need the while loop. If handled returns "true" then just call ContinueDialogAsync() from within the catch block. That way if another error occurs it will be throw out of OnTurnAsync().

Reply:
as agreed offline in response to this comment I have added detailed comment in the code explaining why we have a loop.